week3/README.md

```diff
@@ -0,0 +1,61 @@
+#### Part 1 - Short-answer questions.
+a) 0…5 and 0..<5 are two types of **ranges**. How are they different?</br>
+
+**`0...5` is a *closed range*; it includes both the lower and upper boundaries of the range (in this case: 0, 1, 2, 3, 4, 5).**
+
+**`0..<5` is a *half-open range*; it includes only the lower boundary of the range and not the upper boundary (in this example: 0, 1, 2, 3, 4).**
+
+b) Describe type inference in Swift and give an example.</br>
+**Type inference refers to the Swift compiler's ability to detect the type of a variable or constant based on its value. For example, I could declare a variable named `petAge` and set it to a value of `13`, like so: `var petAge = 13`. Swift would be able to infer that the type of my variable was an `Int`, because I set it to a numeric value with no decimal point.**
+
+c) List three advantages of Playgrounds.
+- **You can set Playground code to auto-run and evaluate and display expression results in real-time (every time you make a change), so you can quickly run experiments with your code (tweaking functions, trying out different input parameter values, etc.)**
+- **Alternatively, you can set your Playground to manually run and just run parts of your code at a time—if you click on the play button to the left of a line you can run your code only up to that line, which is good if you just want to test out a small portion of your code.**
+- **Playgrounds have a Results Sidebar on the righthand side where you can see the results of each line, method and function of your code. You can click on the square show/hide button next to a result to expand it in the main editor section, or the Quick Look button to show the results in a Quick Look display. You can even see Graph or Value History views of the results of a loop.**
+
+d) When does the execution of a while loop end?</br>
+
+**The first time the condition of the while loop is not satisfied. If the condition is never satisfied, it's possible that the while loop never runs.**
+
+e) True or False: Tuples in Swift can contain values of different data types. **True**
+
+f) List three data types you have used in Swift.
+- **String**
+- **Int**
+- **Boolean**
+
+g) To execute alternative code when the condition of an if statement is not met, you can use what clause? **An else clause**
+
+h) What is the third element of the array nums defined below?  **44**
+
+let nums = [5, 0, 44, 20, 1]. 
+
+i) An **expression** is a unit of code that resolves to a single value.
+
+j) Define two ways to unwrap optionals in Swift.</br>
```

Comment:
Forced Unwrapping: This involves adding an exclamation mark (!) after the optional's name. It will return the value inside the optional if one exists. However, if the optional is nil, it will trigger a runtime crash.

Optional Binding: This uses if let or guard let to unwrap an optional. If the optional contains a value, it will be assigned to a new constant (or variable) which you can then use within the scope of the if or guard block.

Nil Coalescing Operator (??): This provides a default value for an optional. If the optional is nil, the default value is used. Otherwise, the value of the optional is used.

Optional Chaining: This lets you call properties, methods, and subscripts on optionals that might be nil. If any link in the chain is nil, the whole chain returns nil gracefully without causing a crash

Reply:
I will have to look into optional chaining, I don’t think I’ve seen that one before.